File: leetcode-java/src/others/easy/itwenty/EmployeeImportance.java
package others.easy.itwenty;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

/**
 * 690 Employee Importance
 */

public class EmployeeImportance {
    public int getImportance(List<Employee> employees, int id) {
        Map<Integer, Employee> map = new HashMap<>();
        for (int i = 0; i < employees.size(); i++) {
            map.put(employees.get(i).id, employees.get(i));
        }

        int[] flag = new int[2000];

        Queue<Employee> queue = new LinkedList<>();
        queue.offer(map.get(id));

        int sum = 0;
        while(!queue.isEmpty()) {
            Employee e = queue.poll();
            flag[e.id] = 1;
            sum += e.importance;
            
            for (int i : e.subordinates) {
                if (flag[i] != 1) {
                    queue.offer(map.get(i));
                }
            }
        }

        return sum;
    }
}

class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};

File: leetcode-scala/src/dfs/easy/EmployeeImportance.scala
package dfs.easy

import tools.Employee

import scala.collection.mutable

/**
 * @package dfs
 * @author 余天堂
 * @create 2020/7/12 16:35
 * @description EmployeeImportance
 */
object EmployeeImportance {

    /**
     * 660 ms   50.00%
     * 50.6 MB  100.00%
     * @param employees
     * @param id
     * @return
     */
    def getImportance(employees: List[Employee], id: Int): Int = {
        var employeeImportance = new mutable.HashMap[Int, Employee]()
        employees.foreach(emp => employeeImportance += (emp.id -> emp))

        def dfs(id: Int): Int = {
            val employee = employeeImportance(id)
            employee.subordinates.foreach(o => employee.importance += dfs(o))
            employee.importance
        }

        dfs(id)
    }
}
